velox/expression/CastExpr.cpp

```diff
@@ -634,81 +659,99 @@ void CastExpr::apply(
     const TypePtr& fromType,
     const TypePtr& toType,
     VectorPtr& result) {
-  LocalDecodedVector decoded(context, *input, rows);
-  auto* rawNulls = decoded->nulls();
-
-  LocalSelectivityVector nonNullRows(*context.execCtx(), rows.end());
-  *nonNullRows = rows;
-  if (rawNulls) {
-    nonNullRows->deselectNulls(rawNulls, rows.begin(), rows.end());
-  }
-
-  LocalSelectivityVector nullRows(*context.execCtx(), rows.end());
-  nullRows->clearAll();
-  if (rawNulls) {
-    *nullRows = rows;
-    nullRows->deselectNonNulls(rawNulls, rows.begin(), rows.end());
+  LocalSelectivityVector newRowsHolder(context);
+  LocalSelectivityVector remainingRowsHolder(context);
+  const SelectivityVector* applyRows = nullptr;
+  const SelectivityVector* remainingRows = nullptr;
+
+  ScopedContextSaver saver;
+  std::vector<VectorPtr> inputValues{input};
+  bool peeled = peelInputEncodings(
+      rows, rows, context, inputValues, true, newRowsHolder, saver);
+  applyRows = peeled ? newRowsHolder.get() : &rows;
+
+  // There may be nulls after peeling, so we deselect null rows.
+  bool deselectNulls = false;
+  if (inputValues[0]->mayHaveNulls()) {
+    LocalDecodedVector decoded(context, *inputValues[0], *applyRows);
+    if (auto* rawNulls = decoded->nulls()) {
+      remainingRowsHolder.get(*applyRows);
+      remainingRowsHolder->deselectNulls(
+          rawNulls, applyRows->begin(), applyRows->end());
+      deselectNulls = true;
```

Comment:
do you see any nulls here? because the part in evaluateInputsAndUpdateRemaningRows() uses decodedVector to remove rows, which should also contain rows from the base vector so the rows passed to peelInputEncodings() should not contain any rows that point to nulls in the peeled base vector